nautobot_device_lifecycle_mgmt/jobs/cve_tracking.py

```diff
@@ -58,3 +72,379 @@ def run(self, published_after, debug=False):  # pylint: disable=too-many-locals,
+                        err,
+                        extra={"object": matching_dlc_cve, "grouping": "CVE Association"},
+                    )
+
+                if str(cve_info["modified_date"][0:10]) != str(matching_dlc_cve.last_modified_date):
+                    self.update_cve(matching_dlc_cve, cve_info)
+
+            # API Rest Timer
+            sleep(self.integration.extra_config.get("api_call_delay", 6))
+
+        self.logger.info(
+            "Performed discovery on all software. Created %s CVE.", cve_counter, extra={"grouping": "CVE Creation"}
+        )
+
+    def create_dlc_cves(self, cpe_cves: dict) -> None:
+        """Create CVE entries in the DLC database.
+
+        Args:
+            cpe_cves (dict): Dictionary of CVEs to be created.
+        """
+        created_count = 0
+
+        for cve, info in cpe_cves.items():
+            try:
+                description = (
+                    f"{info['description'][0:251]}..." if len(info["description"]) > 255 else info["description"]
+                )
+            except TypeError:
+                description = "No Description Provided from NIST DB."
+
+            _, created = CVELCM.objects.get_or_create(
+                name=cve,
+                description=description,
+                published_date=date.fromisoformat(info.get("published_date", "1900-01-01")[0:10]),
+                last_modified_date=date.fromisoformat(info.get("modified_date", "1900-01-01")[0:10]),
+                link=info["url"],
+                cvss=info["cvss_base_score"],
+                severity=info["cvss_severity"],
+                cvss_v2=info["cvssv2_score"],
+                cvss_v3=info["cvssv3_score"],
+                comments="ENTRY CREATED BY NAUTOBOT NIST JOB",
+            )
+
+            if created:
+                created_count += 1
+
+        self.logger.info("Created New CVEs.", extra={"grouping": "CVE Creation"})
+
+    def get_cve_info(self, cpe_software_search_urls: list, software: SoftwareVersion) -> dict:
+        """Search NIST for software and related CVEs.
+
+        Args:
+            cpe_software_search_urls (list): List of URLs to query for CVE information.
+            software (object): Software object being queried.
+
+        Returns:
+            dict: Dictionary containing new and existing CVE information.
+        """
+        for cpe_software_search_url in cpe_software_search_urls:
+            result = self.query_api(cpe_software_search_url)
+
+            all_cve_info = {"new": {}, "existing": {}}
+            if result["totalResults"] > 0:
+                self.logger.info(
+                    "Received %s results.",
+                    result["totalResults"],
+                    extra={"object": software, "grouping": "CVE Creation"},
+                )
+                cve_list = [cve["cve"] for cve in result["vulnerabilities"]]
+                dlc_cves = CVELCM.objects.values_list("name", flat=True)
```

Comment:
Is it possible that some of those keys won't be in the payload? Do we need to validate these exist, perhaps inside of `query_api`?

Reply:
https://csrc.nist.gov/schema/nvd/api/2.0/cve_api_json_2.0.schema

The CVE API response schema states that both "totalResults" and "vulnerabilities" are required return properties with "cve" being a required return property of a returned dictionary in the "vulnerabilities" array.